text/0000-constrained-naked.md

```diff
@@ -0,0 +1,142 @@
+- Feature Name: `constrained_naked`
+- Start Date: 2020-08-06
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+This document attempts to increase the utility of [naked functions](https://github.com/rust-lang/rfcs/blob/master/text/1201-naked-fns.md) by constraining their use and increasing their defined invariants. This document also defines a new calling convention: `extern "custom"`.
+
+# Motivation
+
+Naked functions have long been a feature of compilers. These functions are typically defined as normal functions in every regard, except that the compiler does not emit the function prologue and epilogue. Rust's early attempt to support this feature ([RFC 1201](https://github.com/rust-lang/rfcs/blob/master/text/1201-naked-fns.md)) mostly copied the existing compiler behaviors.
+
+However, naked functions are often avoided in practice because their behavior is not well defined. The root cause of this problem is that naked functions are defined by negation: they are functions which lack a prologue and epilogue. Unfortunately, functions that lack a prologue and epilogue present a number of complicated problems that the compiler needs to solve and developers need to work around. And there is a long history of compilers and developers getting this wrong.
+
+This document seeks to define naked functions in a much more constrained, positivistic way. In doing so, naked functions can become more useful.
+
+# Naked function definition
+
+A naked function is a type of FFI function with a defined calling convention and a body which contains only assembly code which can rely upon the defined calling convention.
+
+A naked function is identified by the `#[naked]` attribute and:
+1. must specify a calling convention besides `extern "Rust"`.
+1. must define only FFI-safe arguments and return types.
+1. must not specify the `#[inline]` or `#[inline(always)]` attribute.
+1. must have a body which contains only a single `asm!()` statement which:
+    1. may be wrapped in an `unsafe` block.
+    1. must not contain any operands except `const` or `sym`.
+    1. must contain the `noreturn` option.
+    1. must not contain any other options except `att_syntax`.
+    1. must ensure that the requirements of the calling convention are followed.
+
+In exchange for the above constraints, the compiler commits to:
+1. produce a clear error if any of the above rules are violated.
+1. never inline the function (implicit `#[inline(never)]`).
+1. emit no instructions to the function body that are not contained in the `asm!()` statement.
```

Comment:
Currently LLVM will insert a `ud2` instruction after the assembly code because of https://github.com/rust-lang/rust/pull/45920. That's not a big issue in practice because we're only adding instructions after the asm rather than before it.

Reply:
I had thought about the `ud2` that is generated because of the `noreturn`. I will add that to the unresolved questions section.